lib/puppet/pops/types/types.rb

```diff
@@ -603,6 +671,111 @@ def _assignable?(o)
       end
     end
 
+    class PIterableType < PAnyType
+      attr_reader :element_type
+
+      def initialize(type)
+        @element_type = type
+      end
+
+      def instance?(o)
+        if @element_type.nil? || @element_type.assignable?(PAnyType::DEFAULT)
+          # Any element_type will do
+          case o
+          when Iterable, String, Hash, Enumerable, Enumerator, Range, PEnumType
+            true
+          when Integer
+            o > 0
+          when PIntegerType
+            o.enumerable?
+          else
+            false
+          end
+        else
+          assignable?(TypeCalculator.infer(o))
+        end
+      end
+
+      def generalize
+        @element_type.nil? ? DEFAULT : PIterableType.new(@element_type.generalize)
+      end
+
+      def hash
+        67 * @element_type.hash
+      end
+
+      def iterable?
+        true
+      end
+
+      def iterable_type
+        self
+      end
+
+      def ==(o)
+        self.class == o.class && @element_type == o.element_type
+      end
+
+      DEFAULT = PIterableType.new(nil)
+
+      protected
+
+      # @api private
+      def _assignable?(o)
+        if @element_type.nil? || @element_type.assignable?(PAnyType::DEFAULT)
+          # Don't request the iterable_type since it might be expensive to compute the
+          # type of its elements
```

Comment:
I got tricked here. Worth adding this to the comment ", since this Iterable accepts Any element, it is enough that o is iterable"